compiler/rustc_builtin_macros/src/assert/context.rs

```diff
@@ -0,0 +1,434 @@
+        cx.expr_block(cx.block(span, stmts))
+    }
+
+    // Initial import to ease the construction of all `core::asserting` statements
+    //
+    // use ::core::asserting::*;
+    fn build_glob_import(&self) -> Stmt {
+        self.cx.stmt_item(
+            self.span,
+            self.cx.item(
+                self.span,
+                Ident::empty(),
+                vec![],
+                ItemKind::Use(UseTree {
+                    prefix: self.cx.path(self.span, self.cx.std_path(&[sym::asserting])),
+                    kind: UseTreeKind::Glob,
+                    span: self.span,
+                }),
+            ),
+        )
+    }
+
+    // The necessary custom `panic!(...)` expression.
+    //
+    // panic!(
+    //     "Assertion failed: ... \n With expansion: ...",
+    //     (&__capture0).printer_kind().new(" ... ", __capture0),
+    //     ...
+    // );
+    fn build_panic(&self, expr_str: &str, fmt_str: &str, panic_path: Path) -> P<Expr> {
+        let initial_tts = [
+            TokenTree::token(
+                token::Literal(token::Lit {
+                    kind: token::LitKind::Str,
+                    symbol: Symbol::intern(&format!(
+                        "Assertion failed: {expr_str}\nWith expansion: {fmt_str}",
+                    )),
+                    suffix: None,
+                }),
+                self.span,
+            ),
+            TokenTree::token(token::Comma, self.span),
+        ];
+
+        let captures_tts = self.captures.iter().flat_map(|cap| {
+            let printer_kind_call = self.cx.expr_method_call(
+                PathSegment {
+                    args: None,
+                    id: DUMMY_NODE_ID,
+                    ident: Ident::new(sym::printer_kind, self.span),
+                },
+                vec![self.cx.expr_paren(
+                    self.span,
+                    self.cx.expr_addr_of(self.span, self.cx.expr_ident(self.span, cap.var)),
+                )],
+                self.span,
+            );
+
+            let new_call = self.cx.expr_method_call(
+                PathSegment {
+                    args: None,
+                    id: DUMMY_NODE_ID,
+                    ident: Ident::new(sym::new, self.span),
+                },
+                vec![
+                    printer_kind_call,
+                    self.cx.expr_lit(
+                        self.span,
+                        LitKind::Str(Symbol::intern(&cap.path_string), StrStyle::Cooked),
+                    ),
+                    self.cx.expr_ident(self.span, cap.var),
+                ],
+                self.span,
+            );
+
+            rustc_parse::nt_to_tokenstream(
+                &Nonterminal::NtExpr(new_call),
+                &self.cx.sess.parse_sess,
+                CanSynthesizeMissingTokens::Yes,
+            )
+            .into_trees()
+            .chain([TokenTree::token(token::Comma, self.span)])
```

Comment:
Please avoid `nt_to_tokenstream` at all cost, especially with `CanSynthesizeMissingTokens::Yes`.

As I understand it's possible to construct the same thing directly as a token stream, so it should be constructed as a token stream, even if it may be less convenient without some kind of `quote!`-like functionality.

Reply:
If possible, I would to keep `nt_to_tokenstream` in `ExprKind::MacCall` to expand `matches!` arguments

Reply:
The "// For macro calls that capture variables. For example, matches!(..)" case?

I'm not sure how it can work (the whole arm, not `nt_to_tokenstream` specifically).
What if the initial input tokens of the macro happen to parse as an expression, but are actually a random token soup?
Also, `parser.parse_expr()` doesn't work speculatively, it may report errors and recover if something goes wrong.

I don't think trying to interpret some other macros' inputs is a good idea.

Reply:
Oh :(

The possible ability of expanding `matches!` arguments or macro arguments in general was one of the main reasons I started this RFC.

* Maybe add a new variant for `MacArgs`? `MacArgs::Matches( ... )`?
* Maybe add a new field in TokenKind::Ident? `TokenKind::Ident(Symbol, /* is_raw */ bool, /* is_path */ bool)` (If `is_path` then try capturing)?

Reply:
It may be possible to reliably detect `matches!`, it's just non-trivial.
- First we need to mark `macro_rules! matches` in the standard library as a lang item, diagnostic item, or whatever, so it's recognizable by the compiler.
- Then we need to support `ExpandResult::Retry` for fn-like built-in macros.
- Then we need to try resolving all macro paths inside `assert!`
    - If the resolution is indeterminate, abandon the whole `assert!` and return `ExpandResult::Retry` for it to be re-scheduled later, this may happen often because all of this happens during macro expansion.
    - If the resolution is determinate and matches the standard `matches`, then it's a success and we can parse the first argument as an expression.

So the main conclusion is, I guess, "not in this PR".

Reply:
Thanks!

If you don't mind, I will probably reach you in the Zulip channel for further clarifications :P